### ISSUE TITLE
tf.nn.moments with tf.concat numerical ambiguity

### ISSUE THREAD

**Issue by hunsteve** [tensorflow/tensorflow#10222]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: custom code, see below
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04 LTE
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: ('v1.1.0-rc0-61-g1ec6ed5', '1.1.0')
- **Bazel version (if compiling from source)**: n/a
- **CUDA/cuDNN version**: CUDA: 8.0 / cuDNN: 5.1
- **GPU model and memory**: GTX960m 4GB and GTX1080 8GB
- **Exact command to reproduce**: run the code below

### Describe the problem
**tf.nn.moments** GPU version produces different mean and variance values for numerically same input tensors. The only difference between the inputs is that they are the outputs of one and two **tf.concat** operations (see the code below). CPU version works well.

The bad output is:
```
input diff:0.0
mean diff:2.98023223877e-08
var diff:7.45058059692e-09
```
The correct output should be:
```
input diff:0.0
mean diff:0.0
var diff:0.0
```

### Source code / logs
```
import numpy as np
import tensorflow as tf

with tf.device("/gpu:0"):
    input = tf.placeholder(shape=[16, 4, 4, 8], dtype=tf.float32)
    input1 = tf.concat([input], axis=0)
    input2 = tf.concat([tf.concat([input], axis=0)], axis=0)

    mean1, var1 = tf.nn.moments(input1, axes=[0,1,2])
    mean2, var2 = tf.nn.moments(input2, axes=[0,1,2])

    input_diff_max = tf.reduce_max(tf.abs(input1 - input2))
    mean_diff_max = tf.reduce_max(tf.abs(mean1 - mean2))
    var_diff_max = tf.reduce_max(tf.abs(var1 - var2))

    with tf.Session() as sess:
        i_v, m_v, v_v = sess.run([input_diff_max, mean_diff_max, var_diff_max], feed_dict={input: np.random.rand(16, 4, 4, 8)})
        print("input diff:{}".format(i_v))
        print("mean diff:{}".format(m_v))
        print("var diff:{}".format(v_v))
```

**Comment by asimshankar** [comment]
@zheng-xq @ekelsen : Mind taking a look?

**Comment by hunsteve** [comment]
Determinism option sounds great. Until then a notice in the docs would be nice, saying that the reduction ops might be non-deterministic.

Currently this behavior does not cause serious problems to us. The way it came up was a test, where we tried to numerically compare the outputs and gradients of our implementation of a method with the reference implementation of that method in tensorflow. We got suprised when the reference implementation was not consistent even with itself in our test.

Shortly after submitting this issue, we found this article: https://www.twosigma.com/insights/a-workaround-for-non-determinism-in-tensorflow
